### DISCUSSION TITLE
Provide hints for kpatch-build errors?

### DISCUSSION THREAD

**Discussion by liu-song-6** [dynup/kpatch#1218] [Ideas]
While testing 0.9.4 earlier today with some random patches (that adds some printk), I got an error like:
```
ERROR: printk.o: 2 unsupported section change(s)
printk.o: changed function: __control_devkmsg
printk.o: changed function: devkmsg_sysctl_set_loglvl
printk.o: changed function: console_unlock
printk.o: changed section .init.text not selected for inclusion
printk.o: changed section .rela.init.text not selected for inclusion
```
This surprised me quite a bit, as I believed we already have the proper support for printk_index. Fortunately, I figured it out soon. It was basically because the random patch added a printk to an __init function. 

This "issue" makes me think whether we can enable kpatch-build to provide hints on common mistake. We can probably use a set of patterns and help messages, and apply them to logger() in kpatch-build. For example, if logger sees a line contains  `changed section .init.text`, it will add a hint `Does any of these patches touch an __init function?`. 

Does this sound like a useful addition to kpatch-build?

**Comment by joe-lawrence** [comment]
Hi @liu-song-6, this reminded me of @divyacote 's kpatch-analyze project https://github.com/dynup/kpatch/pull/1201 .  It may already be prepared to detect changes to functions marked with `__init` and `__exit` annotations without even building.  If you think a static code analyzer for kpatch patches would be worthwhile, you may want to check out that PR.

Also, I think it would be fairly simple to add `kpatch-build` hints as you suggest.   I'm not sure about adding pattern matching directly inside the `logger()` function though -- could that be a maintenance issue if the calling function modifies the string without updating the logger match?  Without seeing all the potential use cases, it may be cleaner to just determine the hint from the place generating the problem, so it can follow up with a "Hint: does build modify an __init function?" msg.
